gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_plugin.rb

```diff
@@ -0,0 +1,191 @@
+agent is running on, where client metrics will be published via UDP.
+      DOCS
+        resolve_client_side_monitoring_port(cfg)
+      end
+
+      option(:client_side_monitoring_publisher,
+        default: ClientSideMonitoring::Publisher.new,
+        docstring: <<-DOCS)
+Allows you to provide a custom client-side monitoring publisher class. By default,
+will use the Client Side Monitoring Agent Publisher.
+      DOCS
+
+      option(:client_side_monitoring_client_id,
+        default: "",
+        docstring: <<-DOCS) do |cfg|
+Allows you to provide an identifier for this client which will be attached to
+all generated client side metrics. Defaults to an empty string.
+        DOCS
+        resolve_client_id(cfg)
+      end
+
+      def add_handlers(handlers, config)
+        if config.client_side_monitoring
+          handlers.add(Handler, step: :initialize)
+          publisher = config.client_side_monitoring_publisher
+          publisher.agent_port = config.client_side_monitoring_port
+        end
+      end
+
+      private
+      def self.resolve_client_side_monitoring_port(cfg)
+        env_source = ENV["AWS_CSM_PORT"]
+        env_source = nil if env_source == ""
+        cfg_source = Aws.shared_config.csm_port(profile: cfg.profile)
+        env_source || cfg_source
+      end
+
+      def self.resolve_client_side_monitoring(cfg)
+        default = true
+        env_source = ENV["AWS_CSM_ENABLED"]
+        env_source = nil if env_source == ""
+        if env_source.is_a?(String) && (env_source.downcase == "false" || env_source.downcase == "f")
+          env_source = false
+        end
+        cfg_source = Aws.shared_config.csm_enabled(profile: cfg.profile)
+        if env_source == false
+          false
+        elsif cfg_source == false
+          false
+        else
+          true
+        end
+      end
+
+      def self.resolve_client_id(cfg)
+        default = ""
+        env_source = ENV["AWS_CSM_CLIENT_ID"]
+        env_source = nil if env_source == ""
+        cfg_source = Aws.shared_config.csm_client_id(profile: cfg.profile)
+        env_source || cfg_source || default
+      end
+
+      class Handler < Seahorse::Client::Handler
+        def call(context)
+          publisher = context.config.client_side_monitoring_publisher
+          service_id = context.config.api.metadata["serviceId"]
+          # serviceId not present in all versions, need a fallback
+          service_id ||= _calculate_service_id(context)
+
+          request_metrics = ClientSideMonitoring::RequestMetrics.new(
+            service: service_id,
+            operation: context.operation.name,
+            client_id: context.config.client_side_monitoring_client_id,
+            timestamp: DateTime.now.strftime('%Q').to_i,
+          )
+          context.metadata[:client_metrics] = request_metrics
+          start_time = Time.now
+          begin
+            resp = @handler.call(context)
+          ensure # Anything different in case of failure?
```

Comment:
Also feel same? as we are tracking those latency data metrics anyway?

Reply:
Tracked in a different plugin, regardless of call success. The spec tests should reflect this.